include/enoki/texture.h

```diff
@@ -667,38 +652,150 @@ template <typename Value, size_t Dimension> class Texture {
+                for (uint32_t iy = 0; iy < Length; iy++) {
+                    pos_i[0][iy * Length + ix] = offset[iy] + pos.x();
+                    pos_i[1][ix * Length + iy] = offset[iy] + pos.y();
+                }
+            }
+        } else if constexpr (Dimension == 3) {
+            constexpr size_t LengthSqr = Length * Length;
+            for (uint32_t ix = 0; ix < Length; ix++) {
+                for (uint32_t iy = 0; iy < Length; iy++) {
+                    for (uint32_t iz = 0; iz < Length; iz++) {
+                        pos_i[0][iz * LengthSqr + iy * Length + ix] =
+                            offset[iz] + pos.x();
+                        pos_i[1][ix * LengthSqr + iz * Length + iy] =
+                            offset[iz] + pos.y();
+                        pos_i[2][iy * LengthSqr + ix * Length + iz] =
+                            offset[iz] + pos.z();
+                    }
+                }
+            }
+        }
+
+        return pos_i;
+    }
+
+    template <typename T> T wrap(const T &pos) const {
+        using Scalar = scalar_t<T>;
+        static_assert(
+                array_size_v<T> == Dimension &&
+                std::is_integral_v<Scalar> &&
+                std::is_signed_v<Scalar>
+        );
+
+        const Array<Int32, Dimension> shape = m_shape_opaque;
+        if (m_wrap_mode == WrapMode::Clamp) {
+            return clamp(pos, 0, shape - 1);
+        } else {
+            const T value_shift_neg = select(pos < 0, pos + 1, pos);
+
+            T div;
+            for (size_t i = 0; i < Dimension; ++i)
+                div[i] = m_inv_resolution[i](value_shift_neg[i]);
+
+            T mod = pos - div * shape;
+            mod[mod < 0] += T(shape);
+
+            if (m_wrap_mode == WrapMode::Mirror)
+                // Starting at 0, flip the texture every other repetition
+                // (flip when: even number of repetitions in negative direction,
+                // or odd number of repetions in positive direction)
+                mod =
+                    select(eq(div & 1, 0) ^ (pos < 0), mod, shape - 1 - mod);
+
+            return mod;
+        }
+    }
+
+    template <typename T>
+    uint32_array_t<value_t<T>> index(const T &pos) const {
+        using Scalar = scalar_t<T>;
+        using Index = uint32_array_t<value_t<T>>;
+        static_assert(
+                array_size_v<T> == Dimension &&
+                std::is_integral_v<Scalar> &&
+                std::is_signed_v<Scalar>
+        );
+
+        Index index;
+        if constexpr (Dimension == 1)
+            index = Index(pos.x());
+        else if constexpr (Dimension == 2)
+            index = Index(
+                fmadd(Index(pos.y()), m_shape_opaque.x(), Index(pos.x())));
+        else if constexpr (Dimension == 3)
+            index = Index(fmadd(
+                fmadd(Index(pos.z()), m_shape_opaque.y(), Index(pos.y())),
+                m_shape_opaque.x(), Index(pos.x())));
```

Comment:
I personnaly find it more readible with the braces ;)
```suggestion
        if constexpr (Dimension == 1) {
            index = Index(pos.x());
        } else if constexpr (Dimension == 2) {
            index = Index(
                fmadd(Index(pos.y()), m_shape_opaque.x(), Index(pos.x())));
        } else if constexpr (Dimension == 3) {
            index = Index(fmadd(
                fmadd(Index(pos.z()), m_shape_opaque.y(), Index(pos.y())),
                m_shape_opaque.x(), Index(pos.x())));
         }
```